File: src/main/java/no/fintlabs/provider/event/request/RequestEventService.java
package no.fintlabs.provider.event.request;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.fintlabs.adapter.models.event.RequestFintEvent;
import no.fintlabs.provider.redis.EventRedisService;
import org.apache.commons.lang3.StringUtils;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.time.LocalDate;
import java.util.Set;
import java.util.concurrent.TimeUnit;

@Slf4j
@Service
@RequiredArgsConstructor
public class RequestEventService {

    private final EventRedisService eventRedisService;

    public void addEvent(RequestFintEvent event) {
        eventRedisService.putRequestFintEvent(event.getCorrId(), event).subscribe(requestAdded -> {
            if (requestAdded) {
                log.info("New RequestFintEvent added: {}", event.getCorrId());
            }
        });
    }

    public void removeEvent(String corrId) {
        eventRedisService.putResponseFintEvent(corrId).subscribe(responseAdded -> {
            if (responseAdded) {
                log.info("New ResponseFintEvent added: {}", corrId);
                eventRedisService.removeRequestFintEvent(corrId).subscribe(deletedAmount -> {
                    if (deletedAmount == 1L) {
                        log.info("Removed RequestFintEvent: {}", corrId);
                    }
                });
            }
        });
    }

    public Flux<RequestFintEvent> getEvents(Set<String> assets, String domainName, String packageName, String resourceName, int size) {
        return eventRedisService.getAllRequestFintEvents()
                .filter(event -> assets.contains(event.getOrgId()))
                .filter(event -> StringUtils.isBlank(domainName) || event.getDomainName().equalsIgnoreCase(domainName))
                .filter(event -> StringUtils.isBlank(packageName) || event.getPackageName().equalsIgnoreCase(packageName))
                .filter(event -> StringUtils.isBlank(resourceName) || event.getResourceName().equalsIgnoreCase(resourceName))
                .take(size > 0 ? size : Long.MAX_VALUE);
    }

    public Mono<RequestFintEvent> getEvent(String corrId) {
        return eventRedisService.getRequestFintEvent(corrId);
    }

}

File: src/main/java/no/fintlabs/provider/event/request/RequestEventService.kt
package no.fintlabs.provider.event.request

import no.fintlabs.adapter.models.event.RequestFintEvent
import no.fintlabs.adapter.models.event.ResponseFintEvent
import no.fintlabs.provider.event.response.ResponseEventTopicProducer
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.*
import java.util.function.Consumer

@Service
class RequestEventService(
    private val requestCache: RequestCache,
    private val responseProducer: ResponseEventTopicProducer
) {

    private val logger = LoggerFactory.getLogger(javaClass)

    init {
        requestCache.onExpired = Consumer { event -> sendExpiredResponse(event) }
    }

    fun getEvents(
        assets: Set<String>,
        domainName: String? = null,
        packageName: String? = null,
        resourceName: String? = null,
        size: Int = 0
    ): List<RequestFintEvent> {
        val stream = requestCache.getAll()
            .filter { assets.contains(it.orgId) }
            .filter { domainName.isNullOrBlank() || it.domainName.equals(domainName, ignoreCase = true) }
            .filter { packageName.isNullOrBlank() || it.packageName.equals(packageName, ignoreCase = true) }
            .filter { resourceName.isNullOrBlank() || it.resourceName.equals(resourceName, ignoreCase = true) }

        return if (size > 0) stream.take(size).toList() else stream.toList()
    }

    fun addEvent(event: RequestFintEvent) {
        if (requestCache.add(event)) {
            logger.debug("Event with corrId: {} added", event.corrId)
        }
    }

    fun removeEvent(corrId: String) {
        requestCache.remove(corrId)
        logger.debug("Event with corrId: {} removed", corrId)
    }

    fun getEvent(corrId: String): Optional<RequestFintEvent> = Optional.ofNullable(requestCache.get(corrId))

    /**
     * Caffeine triggers this automatically when an item expires.
     */
    private fun sendExpiredResponse(request: RequestFintEvent) {
        logger.info("Event {} expired. Sending expired response.", request.corrId)
        responseProducer.sendEvent(request.toResponse(), request)
    }

    private fun RequestFintEvent.toResponse(): ResponseFintEvent =
        ResponseFintEvent().apply {
            corrId = this@toResponse.corrId
            orgId = this@toResponse.orgId
            handledAt = System.currentTimeMillis()
            isFailed = true
            errorMessage = "Event expired."
        }

}